sound/soc/sof/control.c

```diff
@@ -342,24 +345,55 @@ int snd_sof_bytes_ext_get(struct snd_kcontrol *kcontrol,
 	struct snd_sof_dev *sdev = scontrol->sdev;
 	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
 	unsigned int tag = 0;
+	int max_length = SOF_IPC_MSG_MAX_SIZE -
+		sizeof(const struct sof_ipc_ctrl_data) -
+		sizeof(const struct sof_abi_hdr);
+	int err;
+	int ret;
 
-	pm_runtime_get_sync(sdev->dev);
+	ret = pm_runtime_get_sync(sdev->dev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: bytes get failed to resume %d\n",
+			ret);
+		return ret;
+	}
 
-	dev_dbg(sdev->dev, "getting data and command is %d\n", scontrol->cmd);
-	/* get all the mixer data from DSP */
+	/* Decrement size for room for ext bytes header and find out
+	 * the limit from ext bytes size from topology and SOF IPC
+	 * size limit.
+	 */
+	size -= 2 * sizeof(u32);
```

Comment:
what is 2 * sizeof int32_t ?

Reply:
@lgirdwood when sending the bytes from the userspace, the first 2 u32's are actually not part of the actual data to be sent to the firmware. Instead the first uint is a tag and the second is the size of bytes being sent. So we subtract those from the size to be passed to the firmware. 
@singalsu can we please update the comment to reflect it and maybe define a macro like BYTE_CTL_HEADER_SIZE instead of 2?

Reply:
Lets find the C structure that defines this ABI and subtract it's sizeof otherwise this code will break with any ABI update or on another architecture.

Reply:
Yep, I'll search for struct.

I followed the example in sound/soc/intel/skylake/skl-topology.c function skl_tplg_tlv_control_get(). I'm wondering why the size need to be decremented. It prevents getting the max topology defined length of ext bytes data.

Reply:
I searched in ASoC more examples and found one.

Function wm_coeff_tlv_get() in wm_adsp.c does not place the tag & size into header and does not decrement the size.

That way there would be no need for header struct knowledge but I haven't tested yet if that would work.

Reply:
@ranj063 @lgirdwood Based on my quick test the ext bytes get works without placing the header to data[0] and data[1]. I can read back successfully the previously written coefficients to user space. Maybe using this header for get is a misunderstanding.

The header definitely is present in ext bytes put copy_from_user() so this looks different. Any thoughts on this?

Reply:
@singalsu I would follow any prior art that is upstream on the matter. Lets do the same.